### DISCUSSION TITLE
Conan cross-compiling

### DISCUSSION THREAD

**Discussion by avetiso** [mesonbuild/meson#10069] [Q&A] [answered]
Apologies this isn't in the conan repo, they don't have a discussion section and I have received no response in their slack channel for this question :(

I am trying to cross-compile using meson, but executed via the conan helper functions for meson.

We have an internal tool that we use to generate conan profiles based on a set of arguments, so for example `bx mx-conan create realease ios` should call conan with the correct profile and pass it the arch, build type, compiler, OS version, etc.

Using that information in the conanfile.py, I am trying to detect the target OS and provide the corresponding cross file to meson.

```
    def build(self):
        build_args = []

        if self.settings.os == "iOS":
            build_args.append("--cross-file cross/ios.txt")

        meson = Meson(self, build_type=self.settings.build_type)
        meson.configure(build_folder="build")
        meson.build(args=build_args) # pass the cross file here
```

Meson does not seem to do anything with this argument when I pass it to build(), but it doesn't error out, either. I have verified that I am hitting the correct paths I expect to in this code and that the arg is indeed in that list getting passed to build().

If I try to pass it to configure() it does fail, saying that it is an unsupported argument.

The cross file works as expected when I do things manually from the command line.

Is this the idiomatic way to pass cross-compile files to meson when using conan?

**Comment by eli-schwartz** [comment, marked as answer]
I don't know how Conan works, but you will need to make sure that the argument is being passed to whichever Conan function is running the `meson setup` command under the hood.

At a guess, the "build" function seems like it would be running `ninja` after the build had been set up. What's the exact text of the error message when you pass it to configure instead?

**Reply by avetiso** [reply]
The following is the error if I try to pass it to configure(). There are single quotes in the output around the cross-file arg, not sure how I'm supposed to rid myself of those since I create my args as a list of strings and the only entry in that list is the string `'--cross-file cross/ios.txt'`. Perhaps there some simple Python knowledge I don't know lol.

You are right about build() actually deferring to either a call to ninja or 'meson compile' https://github.com/conan-io/conan/blob/develop/conans/client/build/meson.py

Conan does provide a method to simply run meson commands, but when I try to use it to run `setup` there is a -C that's always inserted into the call to meson that ends up breaking it. Seems like a conan issue, not sure why they didn't expose an easy way to use cross files. Doesn't seem to be any serious examples of conan/meson being used together out there for me to study, either.

```
usage: meson [-h] {setup,configure,dist,install,introspect,init,test,wrap,subprojects,help,rewrite,compile,devenv} ...
meson: error: unrecognized arguments: --cross-file cross/ios.txt
base64/1.0.0-5@mx/stable: 
base64/1.0.0-5@mx/stable: ERROR: Package 'b55b855369793ce36245589d74115f074874150a' build failed
base64/1.0.0-5@mx/stable: WARN: Build folder /Users/dario.avetisov/.conan/data/base64/1.0.0-5/mx/stable/build/b55b855369793ce36245589d74115f074874150a
ERROR: base64/1.0.0-5@mx/stable: Error in build() method, line 35
        meson.configure(args=args, build_folder="build")
        ConanException: Error 2 while executing meson "/Users/dario.avetisov/.conan/data/base64/1.0.0-5/mx/stable/build/b55b855369793ce36245589d74115f074874150a" "/Users/dario.avetisov/.conan/data/base64/1.0.0-5/mx/stable/build/b55b855369793ce36245589d74115f074874150a/build" --backend=ninja -Dprefix="/Users/dario.avetisov/.conan/data/base64/1.0.0-5/mx/stable/package/b55b855369793ce36245589d74115f074874150a" -Dlibdir="lib" -Dbindir="bin" -Dsbindir="bin" -Dlibexecdir="bin" -Dincludedir="include" -Ddefault_library="shared" '--cross-file cross/ios.txt' --buildtype=debug
conan create . mx/stable -u -pr ios-debug -s os.version=11.0 -s compiler.version=13.0
Error: Command returned pid 821 exit 1
```

**Reply by avetiso** [reply]
Running the command without the single quotes actually seems to work, so it seems I'm misusing python/conan actually

**Reply by eli-schwartz** [reply]
Alright, looking at that conan definition it seems that `configure(args=...)` accepts a list.

Looking back at your original attempt, the error you get seems to make sense, because you append a single string to the args, and then that entire string gets accepted as a single argument to `meson`. Can you try:

```python
build_args.extend(["--cross-file", "cross/ios.txt"])
```

Then as mentioned above, use it with `meson.configure()` rather than `meson.build()`

**Reply by avetiso** [reply]
that worked! thank you so much, I figured it was the quote that were breaking it, but looks like the single quotes work just fine, it's just that it was within the same string

**Reply by avetiso** [reply]
do you take any sort of donations for your help with FOSS?
